crypto/evp/mac_lib.c

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <stdarg.h>
+#include <openssl/evp.h>
+#include <openssl/err.h>
+#include <openssl/ossl_typ.h>
+#include "internal/nelem.h"
+#include "internal/evp_int.h"
+#include "evp_locl.h"
+
+EVP_MAC_CTX *EVP_MAC_CTX_new_id(int id)
+{
+    const EVP_MAC *mac = EVP_get_macbynid(id);
+
+    if (mac == NULL)
+        return NULL;
+    return EVP_MAC_CTX_new(mac);
+}
+
+EVP_MAC_CTX *EVP_MAC_CTX_new(const EVP_MAC *mac)
+{
+    EVP_MAC_CTX *ctx = OPENSSL_zalloc(sizeof(EVP_MAC_CTX));
+
+    /* Dummy check that there is a freeing function as well */
+    mac->free(NULL);
+
+    if (ctx == NULL
+        || (ctx->data = mac->new()) == NULL) {
+        EVPerr(EVP_F_EVP_MAC_CTX_NEW, ERR_R_MALLOC_FAILURE);
+        OPENSSL_free(ctx);
+        ctx = NULL;
+    } else {
+        ctx->meth = mac;
+    }
+    return ctx;
+}
+
+static void evp_mac_ctx_cleanup(EVP_MAC_CTX *ctx)
+{
+    if (ctx != NULL && ctx->data != NULL) {
+        ctx->meth->free(ctx->data);
+        ctx->data = NULL;
+    }
+}
+
+void EVP_MAC_CTX_free(EVP_MAC_CTX *ctx)
+{
+    evp_mac_ctx_cleanup(ctx);
+    OPENSSL_free(ctx);
+}
+
+int EVP_MAC_CTX_copy(EVP_MAC_CTX *dst, EVP_MAC_CTX *src)
+{
+    /*
+     * NOTE: this is probably wasteful, as this frees a data structure that was
+     * most likely allocated through EVP_MAC_CTX_new() microseconds earlier.
+     */
+    if (dst)
+        dst->meth->free(dst->data);
+
+    *dst = *src;
+
+    if (src->data != NULL && (dst->data = dst->meth->copy(src->data)) == NULL) {
+        EVPerr(EVP_F_EVP_MAC_CTX_COPY, ERR_R_MALLOC_FAILURE);
+        return 0;
```

Comment:
I finally came to realize that this isn't a good implementation, so I'm changing the `copy` EVP_MAC function to behave the same way as other copying functions do.